src/box/txn.c

```diff
@@ -978,6 +978,17 @@ txn_journal_entry_new(struct txn *txn)
 
 		req->approx_len += xrow_approx_len(stmt->row);
 	}
+	if (txn_limbo.is_running_on_split_brain_rollback_triggers) {
+		assert(txn_limbo.is_in_rollback);
+		if (!txn_is_fully_local(txn)) {
+			diag_set(IllegalParams,
+				 "Transactions from the "
+				 "box.ctl.on_replication_split_brain_rollback "
+				 "event trigger must be fully local");
+			return NULL;
+		}
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+	}
```

Comment:
> Given the uncertainty and depth of the problem I would say this patch needs a design before implementation. Is there a document I could read? Otherwise I suggest we make one, to discuss it.

AFAIC, this patch was already designed and discussed in the scope of the repair queue theme: https://www.notion.so/tarantool/WIP-Automatic-handling-of-async-transactions-from-old-terms-in-CE-and-EE-1882542d61a54e69b41c74f96436ad34?pvs=4#065efdadd8bd4ea6a91c13ef9329230f (didn't have time yet to move it to Confluence).

I agree that the requirement to run an arbitrary number of transactions seems too generic and unnecessary. I would prefer having only one transaction available as you described in `Alternative 2`. I don't think that always starting a transaction will introduce a noticeable overhead — moreover, this code is not the critical path.

However, the behavior will be less transparent and case-specific compared to the current solution.

Reply:
/cc @sergepetrenko @Serpentian

Reply:
> I don't think that's really a problem. Why not allow everyone to run fully local transactions during this time?

Because it would mean a bunch of TXNs would slip past the limbo, without being related to the limbo state, but abusing it to ignore the quorum.

Alternatively we could say, that if the limbo is in the rollback state, then it could be treated as "empty", which allows local and async txns to get committed without a quorum indeed. The problem though is what happens if the limbo rollback fails, and some stuff unrelated to the limbo problems was already committed?

That is very edgy. Thanks for the design link btw! Somehow I overlooked in that document, that it would be so intrusive in the code. Specifically the highlighted code chunk that we are discussing. It is a big signal we are doing something wrong, when such an intrusion becomes necessary.

Reply:
> **Alternative 2**
> 
> We start a transaction ourselves unconditionally, force-async. Then user's callback gets executed. The user knows it is already having a txn, and the user's code is supposed to just do whatever operations they need. Then the callback is done, and we try to commit the txn if it is not empty. If the user rolls it back and starts another one inside of the callback, then it won't be force-async, and won't be able to break anything.
> 
> It seems better than the previous solution in a sense that the user doesn't have to leave an open txn hanging. But it is worse in a sense that the user has no choice whether a txn is needed.

I like this variant. It will suit our needs with repair queue just fine and won't introduce any crutches like the current approach. Besides, this trigger is rather low-level, it describes a rarely needed state (similar to `box.ctl.on_schema_init`), so I think this is fine that its semantics differ a bit from other transaction triggers. We could even name this event "internal" to emphasize that it's not intended for general use (although we won't forbid such use, of course).

Besides, we could forbid user-called `box.commit()/box.rollback()` from the trigger similarly to how it was done in de80e0264f7deb58ea86ef85b37b92653a803430

Reply:
I would still recommend to reconsider if this should be an event or a trigger. Shame on me I didn't pay enough attention to it in the document 😣. Events semantics is to notify after the fact. Not to change or prevent or piggyback something. The semantics we need here is clearly a trigger - something that can affect the outcome.

For instance, of `box.status` event callback fails, does it prevent the status from changing? Nope. Then why reacting to this `box.ctl.thisnewlongname` event should change the state of anything?

Another example - `space:on_replace` is not an event, it is a trigger, because it can prevent the replace or add more statements to the same txn. Certainly not an event, right?

Does this make sense?

Reply:
It seems there is some misunderstanding here. From what I see it is implemented like a trigger. So is there a problem with naming or something else?

Sure thing `space:on_replace` is a trigger, but it can be set via
```lua
local trigger = require('trigger')
trigger.set('box.space[512].on_replace', 'my_name.on_replace_trigger', function(...) ... end)
```
And here we refer to `box.space[512].on_replace` as an event upon which a trigger is fired. At least I found this terminology  here: https://github.com/tarantool/doc/issues/3988

Reply:
Aaaaaah, omg, I was looking at `struct event *on_split_brain_rollback` and was thinking why is it an event like for `box.broadcast()`! Sorry, didn't notice it, then all is ok!